snuba/migrations/operations.py

```diff
@@ -101,3 +101,113 @@ def format_sql(self) -> str:
         columns = ", ".join([col.for_schema() for col in self.__columns])
 
         return f"CREATE MATERIALIZED VIEW IF NOT EXISTS {self.__view_name} TO {self.__destination_table_name} ({columns}) AS {self.__query};"
+
+
+class AddColumn(SqlOperation):
+    """
+    Adds a column to a table.
+    """
+
+    def __init__(
+        self,
+        storage_set: StorageSetKey,
+        table_name: str,
+        column: Column,
+        after: Optional[str],
+    ):
+        super().__init__(storage_set)
+        self.__table_name = table_name
+        self.__column = column
+        self.__after = after
+
+    def format_sql(self) -> str:
+        column = self.__column.for_schema()
+        optional_after_clause = (
+            f" AFTER {self.__after}" if self.__after is not None else ""
+        )
+        return f"ALTER TABLE {self.__table_name} ADD COLUMN IF NOT EXISTS {column}{optional_after_clause};"
+
+
+class DropColumn(SqlOperation):
+    """
+    Drops a column from a table.
+
+    You cannot drop a column that is part of the the primary key or the sampling
+    key in the engine expression.
+    """
+
+    def __init__(self, storage_set: StorageSetKey, table_name: str, column_name: str):
+        super().__init__(storage_set)
+        self.__table_name = table_name
+        self.__column_name = column_name
+
+    def format_sql(self) -> str:
+        return f"ALTER TABLE {self.__table_name} DROP COLUMN IF EXISTS {self.__column_name};"
+
+
+class ModifyColumn(SqlOperation):
+    """
+    Modify a column in a table.
+
+    For columns that are included in the primary key, you can only change the type
```

Comment:
We should eventually be able to validate these operations against the current state of the schema, which means we will need to have a current state of the schema we can access (there are probably other reasons we will need to access the current state of the schema, like checking if a column is present during query processing).

I cannot remember whether we landed somewhere on that discussion (whether a list of incremental migration is going to fully replace the schema/DDL for a table).

Reply:
The current state of the schema (essentially what we have now) will most likely exist in a similar format to today for querying and other non DDL purposes. That schema matches the actual database state only if all migrations in the current version have been executed. It could probably tell you if the column was being referenced in code still.

But we don't keep track of the state of the actual database schema and what it should be between every migration that's run. Figuring out whether the column was being referenced or materialized somewhere at that point might be a bit harder.

Reply:
Yeah, this is a good point. One problem is having access to the final state of the schema for the version/commit we are running, which is easy and which is what is needed for everything query processing. A different thing is knowing the expected state at every migration.
As long as we have the final schema, I think we are good.